src/Compiler.cpp

```diff
@@ -201,108 +287,139 @@ TfLiteNode g_nodes[)"
+    // CP: Also if it were written to, we would see a segfault (write to text segment)
+    if (nodes_[i].node.builtin_data || nodes_[i].node.custom_initial_data) {
+      wr << "const_cast<void*>(static_cast<const void*>(&"
+         << prefix_ << "opdata" << i << ")), ";
+    } else {
+      wr << "nullptr, ";
+    }
+    auto regI = nodes_[i].regIndex;
+    if (registrations_[i].code == tflite::BuiltinOperator_CUSTOM) {
+      wr << "OP_" << registrations_[i].custom_name << ", ";
+    }
+    else {
+      wr << "OP_" << tflite::EnumNameBuiltinOperator(registrations_[regI].code) << ", ";
+    }
+    wr << "},\n";
+  }
+  wr << "};";
   // TODO: This code assumes that persistent allocations are made from the end
   // (which is true for the current implementation)
   wr << R"(
-static TfLiteStatus FakeAllocatePersistentBuffer(struct TfLiteContext* ctx,
+static TfLiteStatus AllocatePersistentBuffer(struct TfLiteContext* ctx,
                                                  size_t bytes, void** ptr) {
-  static uint8_t *fakeAllocPtr = g_tensor_arena + sizeof(g_tensor_arena);
+  static uint8_t *AllocPtr = g_tensor_arena + sizeof(g_tensor_arena);
 
-  fakeAllocPtr -= bytes;
-  *ptr = fakeAllocPtr;
+  AllocPtr -= bytes;
+  *ptr = AllocPtr;
   return kTfLiteOk;
 }
 } // namespace
 
-void )"
+TfLiteStatus )"
      << prefix_ << R"(init() {
-  g_ctx.AllocatePersistentBuffer = &FakeAllocatePersistentBuffer;
+  g_ctx.AllocatePersistentBuffer = &AllocatePersistentBuffer;
   g_ctx.tensors = g_tensors;
 )";
   wr << "  g_ctx.tensors_size = " << tensors_.size() << ";\n";
-  for (size_t i = 0; i < tensors_.size(); i++) {
-    auto &t = tensors_[i].tensor;
-    std::string tensorI = "  g_tensors[" + std::to_string(i) + "].";
-    if (t->allocation_type == kTfLiteMmapRo) {
-      wr << tensorI << "data.data = (void*)" << prefix_ << "tensor_data" << i
-         << ";\n";
-    } else {
-      wr << tensorI << "data.data = g_tensor_arena + "
-         << (uintptr_t)t->data.data - (uintptr_t)arena_buf_.data() << ";\n";
-    }
-    wr << tensorI << "type = " << tflmc::to_string(t->type) << ";\n";
-    wr << tensorI << "is_variable = " << t->is_variable << ";\n";
-    wr << tensorI
-       << "allocation_type = " << tflmc::to_string(t->allocation_type) << ";\n";
-    wr << tensorI << "bytes = " << t->bytes << ";\n";
-    wr << tensorI << "dims = (TfLiteIntArray*)&" << prefix_
-       << "tensor_dimension" << i << ";\n";
-    if (t->quantization.type == kTfLiteAffineQuantization) {
-      wr << tensorI << "params.scale = " << t->params.scale << ";\n";
-      wr << tensorI << "params.zero_point = " << t->params.zero_point << ";\n";
-      // TODO: Is this cast safe or does the data need to be non-const?
-      wr << tensorI
-         << "quantization = {kTfLiteAffineQuantization, "
-            "const_cast<void*>(static_cast<const void*>(&"
-         << prefix_ << "quant" << i << "))};\n";
+  // TODO: Do we really support variable tensors?
```

Comment:
do you mean if we can drop the "is_variable" assignment? yes we could, because g_tensors should be .bss zero initialized.

Reply:
I thought about not storing the information in the array, initializing makes perfect sense to me (as some embedded platforms don't zero bss (for good reasons))

Reply:
I would avoid trying to cater to badly non-conforming compilers - at least in the default output. We would need to initialize a lot more if we assume no zero-init. If you have this as use-case it should be an option of the compiler.

Reply:
zeroing bss slows down boot time and thus can be made/is optional on some architectures. I think we should handle this using a memset of the tensor and node array. As some code variation towards visualC compilers (e.g. alignment) is needed somewhen anyway, we could introduce the option to zero variables.

Reply:
see https://github.com/cpetig/tflite_micro_compiler/issues/27